lib/web_ui/lib/src/engine/canvaskit/render_canvas.dart

```diff
@@ -0,0 +1,165 @@
+  void _createNewCanvas(ui.Size physicalSize) {
+    // Clear the container, if it's not empty. We're going to create a new <canvas>.
+    if (canvasElement != null) {
+      canvasElement!.remove();
+    }
+
+    // If `physicalSize` is not precise, use a slightly bigger canvas. This way
+    // we ensure that the rendred picture covers the entire browser window.
+    _pixelWidth = physicalSize.width.ceil();
+    _pixelHeight = physicalSize.height.ceil();
+    final DomCanvasElement htmlCanvas = createDomCanvasElement(
+      width: _pixelWidth,
+      height: _pixelHeight,
+    );
+    canvasElement = htmlCanvas;
+    renderContext = htmlCanvas.contextBitmapRenderer;
+
+    // The DOM elements used to render pictures are used purely to put pixels on
+    // the screen. They have no semantic information. If an assistive technology
+    // attempts to scan picture content it will look like garbage and confuse
+    // users. UI semantics are exported as a separate DOM tree rendered parallel
+    // to pictures.
+    //
+    // Why are layer and scene elements not hidden from ARIA? Because those
+    // elements may contain platform views, and platform views must be
+    // accessible.
+    htmlCanvas.setAttribute('aria-hidden', 'true');
+
+    htmlCanvas.style.position = 'absolute';
+    _updateLogicalHtmlCanvasSize();
+    htmlElement.append(htmlCanvas);
+  }
+
+  /// Ensures that this canvas can draw a frame of the given [size].
+  void ensureSize(ui.Size size) {
+    if (size.isEmpty) {
+      throw CanvasKitError('Cannot create canvases of empty size.');
+    }
+
+    // Check if the frame is the same size as before, and if so, we don't need
+    // to resize the canvas.
+    final ui.Size? previousRenderSize = _currentRenderSize;
+    if (previousRenderSize != null &&
+        size.width == previousRenderSize.width &&
+        size.height == previousRenderSize.height) {
+      // The existing canvas doesn't need to be resized.
+      if (window.devicePixelRatio != _currentDevicePixelRatio) {
+        _updateLogicalHtmlCanvasSize();
+      }
+      return;
+    }
+
+    final ui.Size? previousCanvasSize = _currentCanvasPhysicalSize;
+    // If the canvas is too large or too small, resize it to the exact size of
+    // the frame. We cannot allow the canvas to be larger than the screen
+    // because then when we call `transferFromImageBitmap()` the bitmap will
+    // be scaled to cover the entire canvas.
+    if (previousCanvasSize != null) {
+      canvasElement!.width = size.width.ceilToDouble();
+      canvasElement!.height = size.height.ceilToDouble();
+      _currentCanvasPhysicalSize = size;
+      _pixelWidth = size.width.ceil();
+      _pixelHeight = size.height.ceil();
+      _updateLogicalHtmlCanvasSize();
+    }
+
+    // This is the first frame we have rendered with this canvas.
+    if (_currentCanvasPhysicalSize == null) {
+      _addedToScene = false;
+
+      _createNewCanvas(size);
```

Comment:
Instead of delaying the creation of the canvas, can we create it right away and make it `final` non-null, and then simply update its size? There will be less state to track. We can just update `_currentCanvasPhysicalSize` and move on. Additionally, with the suggestion above, if `RenderCanvas` doesn't need to be responsible for adding itself to the scene host, then we also don't need `_addedToScene`.

Reply:
Done